packages/components/src/internal/components/chart/ChartFieldOption.tsx

```diff
@@ -0,0 +1,230 @@
+import React, { FC, memo, useCallback, useEffect, useMemo, useState } from 'react';
+
+import { SelectInput, SelectInputChange, SelectInputOption } from '../forms/input/SelectInput';
+
+import { QueryModel } from '../../../public/QueryModel/QueryModel';
+
+import { LABKEY_VIS } from '../../constants';
+import { naturalSortByProperty } from '../../../public/sort';
+
+import { OverlayTrigger } from '../../OverlayTrigger';
+import { Popover } from '../../Popover';
+import { RadioGroupInput, RadioGroupOption } from '../forms/input/RadioGroupInput';
+
+import { ChartFieldInfo, ChartTypeInfo } from './ChartBuilderModal';
+import { getFieldDataType } from './utils';
+
+export const getSelectOptions = (
+    model: QueryModel,
+    chartType: ChartTypeInfo,
+    field: ChartFieldInfo
+): SelectInputOption[] => {
+    const allowableTypes = LABKEY_VIS.GenericChartHelper.getAllowableTypes(field);
+
+    return model.queryInfo
+        .getDisplayColumns(model.viewName)
+        .filter(col => {
+            const colType = getFieldDataType(col);
+            const hasMatchingType = allowableTypes.indexOf(colType) > -1;
+            const isMeasureDimensionMatch = LABKEY_VIS.GenericChartHelper.isMeasureDimensionMatch(
+                chartType.name,
+                field,
+                col.measure,
+                col.dimension
+            );
+            return hasMatchingType || isMeasureDimensionMatch;
+        })
+        .sort(naturalSortByProperty('caption'))
+        .map(col => ({ label: col.caption, value: col.fieldKey, data: col }));
+};
+
+export const shouldShowFieldOptions = (field: ChartFieldInfo, selectedType: ChartTypeInfo): boolean => {
+    const showForX = field.name === 'x' && (selectedType.name === 'scatter_plot' || selectedType.name === 'line_plot');
+    const showForY =
+        field.name === 'y' &&
+        (selectedType.name === 'scatter_plot' || selectedType.name === 'line_plot' || selectedType.name === 'box_plot');
+    return showForX || showForY;
+};
+
+const SCALE_TRANS_TYPES = [
+    { value: 'linear', label: 'Linear' },
+    { value: 'log', label: 'Log' },
+];
+
+const SCALE_RANGE_TYPES = [
+    { value: 'automatic', label: 'Automatic' },
+    { value: 'manual', label: 'Manual' },
+];
+
+const DEFAULT_SCALE_VALUES = { type: 'automatic', trans: 'linear' };
+
+interface ChartFieldOptionProps {
+    field: ChartFieldInfo;
+    fieldValue?: SelectInputOption;
+    model: QueryModel;
+    onScaleChange: (field: string, key: string, value: string | number, reset?: boolean) => void;
+    onSelectFieldChange: (name: string, value: any, selectedOption: SelectInputOption) => void;
```

Comment:
I think it is safe to type `value` as `string` here because `getSelectOptions` is using fieldKey as the value.

Reply:
done